--- a/drivers/media/dvb/dvb-usb/dvb-usb-ids.h
+++ b/drivers/media/dvb/dvb-usb/dvb-usb-ids.h
@@ -358,4 +358,5 @@
 #define USB_PID_TECHNISAT_USB2_HDCI_V2			0x0002
 #define USB_PID_TECHNISAT_AIRSTAR_TELESTICK_2		0x0004
 #define USB_PID_TECHNISAT_USB2_DVB_S2			0x0500
+#define USB_PID_AUGUST_DVBT205				0xa803
 #endif
--- a/drivers/media/dvb/dvb-usb/rtl28xxu.c
+++ b/drivers/media/dvb/dvb-usb/rtl28xxu.c
@@ -1157,6 +1157,7 @@ enum rtl28xxu_usb_table_entry {
 	RTL2832U_0CCD_00A9,
 	RTL2832U_1F4D_B803,
 	RTL2832U_0CCD_00B3,
+	RTL2838U_1F4D_A803,
 };
 
 static struct usb_device_id rtl28xxu_table[] = {
@@ -1175,6 +1176,10 @@ static struct usb_device_id rtl28xxu_table[] = {
 		USB_DEVICE(USB_VID_GTEK, USB_PID_DELOCK_USB2_DVBT)},
 	[RTL2832U_0CCD_00B3] = {
 		USB_DEVICE(USB_VID_TERRATEC, USB_PID_NOXON_DAB_STICK)},
+
+	/* RTL2838U */
+	[RTL2838U_1F4D_A803] {
+		USB_DEVICE(USB_VID_GTEK, USB_PID_AUGUST_DVBT205)},
 	{} /* terminating entry */
 };
 
@@ -1288,7 +1293,7 @@ static struct dvb_usb_device_properties rtl28xxu_properties[] = {
 
 		.i2c_algo = &rtl28xxu_i2c_algo,
 
-		.num_device_descs = 3,
+		.num_device_descs = 4,
 		.devices = {
 			{
 				.name = "Terratec Cinergy T Stick Black",
@@ -1308,6 +1313,12 @@ static struct dvb_usb_device_properties rtl28xxu_properties[] = {
 					&rtl28xxu_table[RTL2832U_0CCD_00B3],
 				},
 			},
+			{
+				.name = "AUGUST DVB-T205",
+				.warm_ids = {
+					&rtl28xxu_table[RTL2838U_1F4D_A803],
+				},
+			},
 		}
 	},
 
