From 413f07e7278d43986182207f01f3a22d973fbc54 Mon Sep 17 00:00:00 2001
From: Dom Cobley <dc4@broadcom.com>
Date: Mon, 26 Mar 2012 22:18:37 +0100
Subject: [PATCH] Cast to avoid warning


diff --git a/drivers/video/bcm2708_fb.c b/drivers/video/bcm2708_fb.c
index f23c2d7..fc4e89a 100644
--- a/drivers/video/bcm2708_fb.c
+++ b/drivers/video/bcm2708_fb.c
@@ -191,7 +191,7 @@ static int bcm2708_fb_check_var(struct fb_var_screeninfo *var,
 
 static int bcm2708_fb_set_par(struct fb_info *info)
 {
-	unsigned val = 0;
+	uint32_t val = 0;
 	struct bcm2708_fb *fb = to_bcm2708(info);
 	volatile struct fbinfo_s *fbinfo = fb->info;
 	fbinfo->xres = info->var.xres;
@@ -243,7 +243,7 @@ static int bcm2708_fb_set_par(struct fb_info *info)
 	    ("BCM2708FB: start = %p,%p width=%d, height=%d, bpp=%d, pitch=%d size=%d success=%d\n",
 	     (void *)fb->fb.screen_base, (void *)fb->fb.fix.smem_start,
 	     fbinfo->xres, fbinfo->yres, fbinfo->bpp,
-	     fbinfo->pitch, fb->fb.screen_size, val);
+	     fbinfo->pitch, (int)fb->fb.screen_size, val);
 
 	return val;
 }
