From b486ac9cc0b30f0a1ee191e42c8a7580142b516a Mon Sep 17 00:00:00 2001
From: popcornmix <popcornmix@gmail.com>
Date: Mon, 16 Apr 2012 12:55:39 +0100
Subject: [PATCH] Fix harmless base/size typo


diff --git a/arch/arm/mach-bcm2708/vc_mem.c b/arch/arm/mach-bcm2708/vc_mem.c
index c994374..67c10ee 100644
--- a/arch/arm/mach-bcm2708/vc_mem.c
+++ b/arch/arm/mach-bcm2708/vc_mem.c
@@ -208,14 +208,14 @@ vc_mem_ioctl(struct file *file, unsigned int cmd, unsigned long arg)
 		}
 	case VC_MEM_IOC_MEM_BASE:
 		{
-			// Get the videocore memory size first
+			// Get the videocore memory base
 			vc_mem_get_base();
 
 			LOG_DBG("%s: VC_MEM_IOC_MEM_BASE=%u", __func__,
-				mm_vc_mem_size);
+				mm_vc_mem_base);
 
-			if (copy_to_user((void *) arg, &mm_vc_mem_size,
-					 sizeof (mm_vc_mem_size)) != 0) {
+			if (copy_to_user((void *) arg, &mm_vc_mem_base,
+					 sizeof (mm_vc_mem_base)) != 0) {
 				rc = -EFAULT;
 			}
 			break;
