From 906501202a38d6d9697cc7024928ec9d93d37d8d Mon Sep 17 00:00:00 2001
From: popcornmix <popcornmix@gmail.com>
Date: Tue, 8 May 2012 23:12:13 +0100
Subject: [PATCH] possible fix for sdcard missing status. Thank naren


Index: linux-3.3.8/drivers/mmc/host/sdhci-bcm2708.c
===================================================================
--- linux-3.3.8.orig/drivers/mmc/host/sdhci-bcm2708.c	2012-08-30 21:52:51.279382575 +0200
+++ linux-3.3.8/drivers/mmc/host/sdhci-bcm2708.c	2012-08-30 21:53:19.659523309 +0200
@@ -1291,6 +1291,14 @@
         return 1;
 }
 
+static unsigned int sdhci_bcm2708_missing_status(struct sdhci_host *host)
+{
+	if(host->last_cmdop == MMC_SEND_STATUS)
+		return 1;
+	else
+		return 0;
+}
+
 /***************************************************************************** \
  *									     *
  * Device ops								     *
@@ -1328,6 +1336,7 @@
 	.spurious_crc_acmd51 = sdhci_bcm2708_quirk_spurious_crc,
 	.voltage_broken = sdhci_bcm2708_quirk_voltage_broken,
 	.uhs_broken = sdhci_bcm2708_uhs_broken,
+	.missing_status = sdhci_bcm2708_missing_status,
 };
 
 /*****************************************************************************\
Index: linux-3.3.8/drivers/mmc/host/sdhci.c
===================================================================
--- linux-3.3.8.orig/drivers/mmc/host/sdhci.c	2012-08-30 21:52:11.623185933 +0200
+++ linux-3.3.8/drivers/mmc/host/sdhci.c	2012-08-30 21:53:19.663523367 +0200
@@ -974,6 +974,11 @@
 	if ((cmd->data != NULL) || (cmd->flags & MMC_RSP_BUSY))
 		mask |= SDHCI_DATA_INHIBIT;
 
+	if(host->ops->missing_status && (cmd->opcode == MMC_SEND_STATUS)) {
+		timeout = 100;
+		mask |= SDHCI_DATA_INHIBIT;
+	}
+
 	/* We shouldn't wait for data inihibit for stop commands, even
 	   though they might use busy signaling */
 	if (host->mrq->data && (cmd == host->mrq->data->stop))
Index: linux-3.3.8/drivers/mmc/host/sdhci.h
===================================================================
--- linux-3.3.8.orig/drivers/mmc/host/sdhci.h	2012-08-30 21:51:00.022830892 +0200
+++ linux-3.3.8/drivers/mmc/host/sdhci.h	2012-08-30 21:53:19.663523367 +0200
@@ -291,6 +291,7 @@
 	unsigned int	(*spurious_crc_acmd51)(struct sdhci_host *host);
 	unsigned int	(*voltage_broken)(struct sdhci_host *host);
 	unsigned int	(*uhs_broken)(struct sdhci_host *host);
+	unsigned int	(*missing_status)(struct sdhci_host *host);
 
 	void	(*hw_reset)(struct sdhci_host *host);
 };
