From d461c6d577070637ceeae52ea019e6d907b21c47 Mon Sep 17 00:00:00 2001
From: popcornmix <popcornmix@gmail.com>
Date: Wed, 11 Jul 2012 13:55:11 +0100
Subject: [PATCH] Add missing UART0_CLOCK from last commit


Index: linux-3.3.8/arch/arm/mach-bcm2708/bcm2708.c
===================================================================
--- linux-3.3.8.orig/arch/arm/mach-bcm2708/bcm2708.c	2012-08-30 21:33:12.725538446 +0200
+++ linux-3.3.8/arch/arm/mach-bcm2708/bcm2708.c	2012-08-30 21:41:27.795993368 +0200
@@ -182,7 +182,7 @@
  * These are fixed clocks.
  */
 static struct clk ref24_clk = {
-	.rate = 3000000,	/* The UART is clocked at 3MHz via APB_CLK */
+	.rate = UART0_CLOCK,	/* The UART is clocked at 3MHz via APB_CLK */
 };
 
 static struct clk osc_clk = {
Index: linux-3.3.8/arch/arm/mach-bcm2708/include/mach/platform.h
===================================================================
--- linux-3.3.8.orig/arch/arm/mach-bcm2708/include/mach/platform.h	2012-08-30 21:33:12.725538446 +0200
+++ linux-3.3.8/arch/arm/mach-bcm2708/include/mach/platform.h	2012-08-30 21:41:27.795993368 +0200
@@ -216,7 +216,7 @@
 #define PM_RSTS_HADDRF_SET                                 0x00000002
 #define PM_RSTS_HADDRQ_SET                                 0x00000001
 
-
+#define UART0_CLOCK      3000000
 
 #endif
 
