From e76b20accf3be6b4ddd63dee9313771bdd8ff4b1 Mon Sep 17 00:00:00 2001
From: popcornmix <popcornmix@gmail.com>
Date: Tue, 24 Jul 2012 11:11:03 +0100
Subject: [PATCH] Fix var.width/var.height. They actually mean display size.
 See #65


diff --git a/drivers/video/bcm2708_fb.c b/drivers/video/bcm2708_fb.c
index f2c55a6..5ca8aca 100644
--- a/drivers/video/bcm2708_fb.c
+++ b/drivers/video/bcm2708_fb.c
@@ -374,8 +374,8 @@ static int bcm2708_fb_register(struct bcm2708_fb *fb)
 	fb->fb.var.vmode = FB_VMODE_NONINTERLACED;
 	fb->fb.var.activate = FB_ACTIVATE_NOW;
 	fb->fb.var.nonstd = 0;
-	fb->fb.var.height = fbwidth;
-	fb->fb.var.width = fbheight;
+	fb->fb.var.height = -1;		/* height of picture in mm    */
+	fb->fb.var.width = -1;		/* width of picture in mm    */
 	fb->fb.var.accel_flags = 0;
 
 	fb->fb.monspecs.hfmin = 0;
