Index: linux-3.3.8/arch/arm/mach-bcm2708/bcm2708.c
===================================================================
--- linux-3.3.8.orig/arch/arm/mach-bcm2708/bcm2708.c	2012-09-16 19:22:59.000000000 +0200
+++ linux-3.3.8/arch/arm/mach-bcm2708/bcm2708.c	2012-09-16 19:31:40.419948507 +0200
@@ -69,8 +69,6 @@
  */
 #define DMA_MASK_BITS_COMMON 32
 
-static DEFINE_CLOCK_DATA(cd);
-
 /* command line parameters */
 static unsigned boardrev, serial;
 
@@ -178,17 +176,9 @@
 	}
 }
 
-unsigned long long sched_clock(void)
-{
-	u32 cyc = clocksource_stc.read(&clocksource_stc);
-	return cyc_to_fixed_sched_clock(&cd, cyc, clocksource_stc.mask,
-			SC_MULT, SC_SHIFT);
-}
-
-static void notrace bcm2708_update_sched_clock(void)
+static u32 bcm2708_read_sched_clock(void)
 {
-	u32 cyc = clocksource_stc.read(&clocksource_stc);
-	update_sched_clock(&cd, cyc, clocksource_stc.mask);
+	return clocksource_stc.read(&clocksource_stc);
 }
 
 /*
@@ -708,8 +698,7 @@
 	 */
 	setup_irq(IRQ_TIMER3, &bcm2708_timer_irq);
 
-	init_fixed_sched_clock(&cd, bcm2708_update_sched_clock, 32,
-		STC_FREQ_HZ, SC_MULT, SC_SHIFT);
+	setup_sched_clock(bcm2708_read_sched_clock, 32, STC_FREQ_HZ);
 
 	timer0_clockevent.mult =
 	    div_sc(STC_FREQ_HZ, NSEC_PER_SEC, timer0_clockevent.shift);
