From 04e71eb30083d387cdaabb046bedb3174999f7cd Mon Sep 17 00:00:00 2001
From: Dom Cobley <dc4@broadcom.com>
Date: Tue, 31 Jul 2012 16:42:03 +0100
Subject: [PATCH] Sync with bootc's file


diff --git a/arch/arm/mach-bcm2708/power.c b/arch/arm/mach-bcm2708/power.c
index d147bb6..256bf1a 100644
--- a/arch/arm/mach-bcm2708/power.c
+++ b/arch/arm/mach-bcm2708/power.c
@@ -14,7 +14,6 @@
 #include <linux/module.h>
 #include <linux/semaphore.h>
 #include <linux/bug.h>
-#include <linux/delay.h>
 #include <mach/power.h>
 #include <mach/vcio.h>
 #include <mach/arm_power.h>
@@ -97,6 +96,7 @@ int bcm_power_request(BCM_POWER_HANDLE_T handle, uint32_t request)
 				bcm_mailbox_write(MBOX_CHAN_POWER,
 						  global_request << 4);
 
+				/* Wait for a response during power-up */
 				if (global_request & ~g_state.global_request) {
 					rc = bcm_mailbox_read(MBOX_CHAN_POWER,
 							      &actual);
@@ -111,14 +111,14 @@ int bcm_power_request(BCM_POWER_HANDLE_T handle, uint32_t request)
 
 				if (rc == 0) {
 					if (actual != global_request) {
-						printk(KERN_INFO
-						     "%s: Fail: prev global %x, new global %x, actual %x request %x, others_request %x\n",
+						printk(KERN_ERR
+						     "%s: prev global %x, new global %x, actual %x, request %x, others_request %x\n",
 						     __func__,
 						     g_state.global_request,
 						     global_request, actual, request, others_request);
 						/* A failure */
-					//	BUG_ON((others_request & actual)
-					//	       != others_request);
+						BUG_ON((others_request & actual)
+						       != others_request);
 						request &= actual;
 						rc = -EIO;
 					}
