From cbc60c791cb6fafea5ddae0d8429ac0923237128 Mon Sep 17 00:00:00 2001
From: popcornmix <popcornmix@gmail.com>
Date: Tue, 21 Aug 2012 12:25:47 +0100
Subject: [PATCH] Another fix for 10 second hang on closing sound driver


diff --git a/sound/arm/bcm2835-pcm.c b/sound/arm/bcm2835-pcm.c
index 6cc80dd..bc146a2 100755
--- a/sound/arm/bcm2835-pcm.c
+++ b/sound/arm/bcm2835-pcm.c
@@ -72,7 +72,7 @@ static irqreturn_t bcm2835_playback_fifo_irq(int irq, void *dev_id)
 			  frames_to_bytes(alsa_stream->substream->runtime, alsa_stream->substream->runtime->status->hw_ptr),
 			  new_period);
 	if (alsa_stream->buffer_size) {
-		alsa_stream->pos += consumed;
+		alsa_stream->pos += consumed &~ (1<<30);
 		alsa_stream->pos %= alsa_stream->buffer_size;
 	}
 
diff --git a/sound/arm/bcm2835.h b/sound/arm/bcm2835.h
index 3db5e51..b966e28 100755
--- a/sound/arm/bcm2835.h
+++ b/sound/arm/bcm2835.h
@@ -28,7 +28,7 @@
 /*
 #define AUDIO_DEBUG_ENABLE
 #define AUDIO_VERBOSE_DEBUG_ENABLE
-*.
+*/
 
 /* Debug macros */
 
