From a915a4f6e8969a599627ca6b3e2cd7714608d222 Mon Sep 17 00:00:00 2001
From: Frank Buss <fb@frank-buss.de>
Date: Thu, 23 Aug 2012 02:42:34 +0200
Subject: [PATCH] enabling the realtime clock 1-wire chip DS1307 and 1-wire on
 GPIO4 (as a module)


Index: linux-3.3.8/arch/arm/configs/bcmrpi_defconfig
===================================================================
--- linux-3.3.8.orig/arch/arm/configs/bcmrpi_defconfig	2012-09-19 11:27:02.000000000 +0200
+++ linux-3.3.8/arch/arm/configs/bcmrpi_defconfig	2012-09-19 11:27:43.376026243 +0200
@@ -738,6 +738,8 @@
 CONFIG_LEDS_TRIGGER_TIMER=m
 CONFIG_LEDS_TRIGGER_HEARTBEAT=m
 CONFIG_LEDS_TRIGGER_DEFAULT_ON=m
+CONFIG_RTC_CLASS=y
+CONFIG_RTC_DRV_DS1307=m
 CONFIG_UIO=m
 CONFIG_UIO_PDRV=m
 CONFIG_UIO_PDRV_GENIRQ=m
Index: linux-3.3.8/arch/arm/mach-bcm2708/bcm2708.c
===================================================================
--- linux-3.3.8.orig/arch/arm/mach-bcm2708/bcm2708.c	2012-09-19 11:27:01.000000000 +0200
+++ linux-3.3.8/arch/arm/mach-bcm2708/bcm2708.c	2012-09-19 11:29:04.748429748 +0200
@@ -32,6 +32,7 @@
 #include <linux/io.h>
 #include <linux/module.h>
 #include <linux/spi/spi.h>
+#include <linux/w1-gpio.h>
 
 #include <linux/version.h>
 #include <linux/clkdev.h>
@@ -68,6 +69,9 @@
  * more legitimate.
  */
 #define DMA_MASK_BITS_COMMON 32
+ 
+// use GPIO 4 for the one-wire GPIO pin, if enabled
+#define W1_GPIO 4
 
 /* command line parameters */
 static unsigned boardrev, serial;
@@ -281,6 +285,19 @@
 };
 #endif /* CONFIG_MMC_BCM2708 */
 
+#if defined(CONFIG_W1_MASTER_GPIO) || defined(CONFIG_W1_MASTER_GPIO_MODULE)
+static struct w1_gpio_platform_data w1_gpio_pdata = {
+	.pin = W1_GPIO,
+	.is_open_drain = 0,
+};
+
+static struct platform_device w1_device = {
+	.name = "w1-gpio",
+	.id = -1,
+	.dev.platform_data = &w1_gpio_pdata,
+};
+#endif
+
 static u64 fb_dmamask = DMA_BIT_MASK(DMA_MASK_BITS_COMMON);
 
 static struct platform_device bcm2708_fb_device = {
@@ -582,6 +599,9 @@
 #ifdef CONFIG_BCM2708_GPIO
 	bcm_register_device(&bcm2708_gpio_device);
 #endif
+#if defined(CONFIG_W1_MASTER_GPIO) || defined(CONFIG_W1_MASTER_GPIO_MODULE)
+	platform_device_register(&w1_device);
+#endif
 	bcm_register_device(&bcm2708_systemtimer_device);
 #ifdef CONFIG_MMC_BCM2708
 	bcm_register_device(&bcm2708_mci_device);
