From afe58ac9e71636ff2f46e231c9d5fcd615ef933d Mon Sep 17 00:00:00 2001
From: popcornmix <popcornmix@gmail.com>
Date: Fri, 24 Aug 2012 23:25:13 +0100
Subject: [PATCH] Turn on microframe_schedule by default. Can still be
 disabled on command line


diff --git a/drivers/usb/host/dwc_otg/dwc_otg_driver.c b/drivers/usb/host/dwc_otg/dwc_otg_driver.c
index 8ddc9ae..e7f99e1 100644
--- a/drivers/usb/host/dwc_otg/dwc_otg_driver.c
+++ b/drivers/usb/host/dwc_otg/dwc_otg_driver.c
@@ -60,7 +60,7 @@
 #define DWC_DRIVER_VERSION	"3.00a 10-AUG-2012"
 #define DWC_DRIVER_DESC		"HS OTG USB Controller driver"
 
-bool microframe_schedule;
+bool microframe_schedule=true;
 
 static const char dwc_driver_name[] = "dwc_otg";
 
