From 87d5bc10fb0876be295314213ef4478a94c0e7bb Mon Sep 17 00:00:00 2001
From: popcornmix <popcornmix@gmail.com>
Date: Tue, 11 Sep 2012 00:58:31 +0100
Subject: [PATCH] Fix for KALLSYMS_EXTRA_PASS requirement. Thanks asb for
 finding this.


Index: linux-3.3.8/arch/arm/kernel/vmlinux.lds.S
===================================================================
--- linux-3.3.8.orig/arch/arm/kernel/vmlinux.lds.S	2012-06-01 09:16:13.000000000 +0200
+++ linux-3.3.8/arch/arm/kernel/vmlinux.lds.S	2012-09-19 15:04:51.688630150 +0200
@@ -183,7 +183,9 @@
 	}
 #endif
 
+#ifdef CONFIG_SMP
 	PERCPU_SECTION(L1_CACHE_BYTES)
+#endif
 
 #ifdef CONFIG_XIP_KERNEL
 	__data_loc = ALIGN(4);		/* location in binary */
