From 0875260c9a74e5b46a1a5052dd8472cfec8d1d22 Mon Sep 17 00:00:00 2001
From: popcornmix <popcornmix@gmail.com>
Date: Thu, 20 Sep 2012 21:45:25 +0100
Subject: [PATCH] Enable multiple ALSA channels


Index: linux-3.3.8/arch/arm/mach-bcm2708/bcm2708.c
===================================================================
--- linux-3.3.8.orig/arch/arm/mach-bcm2708/bcm2708.c	2012-09-26 09:41:04.200764544 +0200
+++ linux-3.3.8/arch/arm/mach-bcm2708/bcm2708.c	2012-09-26 09:41:09.312789884 +0200
@@ -506,6 +506,48 @@
 	       .resource = 0,
 	       .num_resources = 0,
 	       },
+	[1] = {
+	       .name = "bcm2835_AUD1",
+	       .id = 1,		/* second audio device */
+	       .resource = 0,
+	       .num_resources = 0,
+	       },
+	[2] = {
+	       .name = "bcm2835_AUD2",
+	       .id = 2,		/* third audio device */
+	       .resource = 0,
+	       .num_resources = 0,
+	       },
+	[3] = {
+	       .name = "bcm2835_AUD3",
+	       .id = 3,		/* forth audio device */
+	       .resource = 0,
+	       .num_resources = 0,
+	       },
+	[4] = {
+	       .name = "bcm2835_AUD4",
+	       .id = 4,		/* fifth audio device */
+	       .resource = 0,
+	       .num_resources = 0,
+	       },
+	[5] = {
+	       .name = "bcm2835_AUD5",
+	       .id = 5,		/* sixth audio device */
+	       .resource = 0,
+	       .num_resources = 0,
+	       },
+	[6] = {
+	       .name = "bcm2835_AUD6",
+	       .id = 6,		/* seventh audio device */
+	       .resource = 0,
+	       .num_resources = 0,
+	       },
+	[7] = {
+	       .name = "bcm2835_AUD7",
+	       .id = 7,		/* eighth audio device */
+	       .resource = 0,
+	       .num_resources = 0,
+	       },
 };
 
 static struct resource bcm2708_spi_resources[] = {
