From 965b922076edccc530bd2a05e2fbabc1dd8b06a8 Mon Sep 17 00:00:00 2001
From: popcornmix <popcornmix@gmail.com>
Date: Sat, 22 Sep 2012 18:56:57 +0100
Subject: [PATCH] Fix Kernel OOPS in SDHCI on boot with the llm patch. Thanks
 ddv2005


Index: linux-3.3.8/drivers/mmc/host/sdhci.c
===================================================================
--- linux-3.3.8.orig/drivers/mmc/host/sdhci.c	2012-09-26 09:40:40.660647814 +0200
+++ linux-3.3.8/drivers/mmc/host/sdhci.c	2012-09-26 09:42:03.701059581 +0200
@@ -144,9 +144,9 @@
 	if(enable_llm)
 	{
 		local_irq_disable();
-		enable_irq(host->irq);
 		if(host->second_irq)
 			enable_irq(host->second_irq);
+		enable_irq(host->irq);
 	}
 #endif
 	spin_unlock(&host->lock);
@@ -178,9 +178,9 @@
 	if(enable_llm)
 	{
 		local_irq_disable();
-		enable_irq(host->irq);
 		if(host->second_irq)
 			enable_irq(host->second_irq);
+		enable_irq(host->irq);
 	}
 #endif
 	spin_unlock_irqrestore(&host->lock,flags);
@@ -294,9 +294,7 @@
 			return;
 		}
 		timeout--;
-		sdhci_spin_enable_schedule(host);
 		mdelay(1);
-		sdhci_spin_disable_schedule(host);
 	}
 
 	if (host->ops->platform_reset_exit)
