From 8bccc7a0340404cc14f831dd9c06d3528b5ebcff Mon Sep 17 00:00:00 2001
From: popcornmix <popcornmix@gmail.com>
Date: Sat, 6 Oct 2012 16:18:27 +0100
Subject: [PATCH] Remove BUG() call from vchiq that is trigger when alsa is
 killed


diff --git a/drivers/misc/vc04_services/interface/vchiq_arm/vchiq_core.c b/drivers/misc/vc04_services/interface/vchiq_arm/vchiq_core.c
index 0b84988..b8675d8 100644
--- a/drivers/misc/vc04_services/interface/vchiq_arm/vchiq_core.c
+++ b/drivers/misc/vc04_services/interface/vchiq_arm/vchiq_core.c
@@ -3180,7 +3180,7 @@ vchiq_bulk_transmit(VCHIQ_SERVICE_HANDLE_T handle, const void *data, int size,
 		mode, VCHIQ_BULK_TRANSMIT);
 
 	/* This call is for kernel thread use and should not be interrupted */
-	BUG_ON(status == VCHIQ_RETRY);
+	// dc4: remove as it does happen: BUG_ON(status == VCHIQ_RETRY);
 	return status;
 }
 
