From 7aa8d823bd9feb758c2e0755d7d3589481bf6b36 Mon Sep 17 00:00:00 2001
From: Stephan Raue <stephan@openelec.tv>
Date: Mon, 8 Oct 2012 16:33:17 +0200
Subject: [PATCH] fix build with CONFIG_BCM2708_VCHIQ is disabled


Index: linux-3.3.8/drivers/misc/Makefile
===================================================================
--- linux-3.3.8.orig/drivers/misc/Makefile	2012-10-10 17:59:32.000000000 +0200
+++ linux-3.3.8/drivers/misc/Makefile	2012-10-10 18:30:41.028352691 +0200
@@ -49,5 +49,5 @@
 obj-$(CONFIG_USB_SWITCH_FSA9480) += fsa9480.o
 obj-$(CONFIG_ALTERA_STAPL)	+=altera-stapl/
 obj-$(CONFIG_MAX8997_MUIC)	+= max8997-muic.o
-obj-y                          += vc04_services/
+obj-$(CONFIG_BCM2708_VCHIQ) += vc04_services/
 
