From cdbb7932058639af2114a709380fb0d86d2964b7 Mon Sep 17 00:00:00 2001
From: popcornmix <popcornmix@gmail.com>
Date: Thu, 12 Apr 2012 12:16:31 +0100
Subject: [PATCH] add temporary workaround for fbset crashes


diff --git a/drivers/video/bcm2708_fb.c b/drivers/video/bcm2708_fb.c
index fc4e89a..5b27287 100644
--- a/drivers/video/bcm2708_fb.c
+++ b/drivers/video/bcm2708_fb.c
@@ -215,6 +215,8 @@ static int bcm2708_fb_set_par(struct fb_info *info)
 	/* inform vc about new framebuffer */
 	bcm_mailbox_write(MBOX_CHAN_FB, fb->dma);
 
+	/* workaround occasional failure to read results. TODO: replace fb driver with vchiq version */
+	msleep(10);
 	/* wait for response */
 	bcm_mailbox_read(MBOX_CHAN_FB, &val);
 
