From 2466e92ca68c25220681173a99154330e0e41c2c Mon Sep 17 00:00:00 2001
From: popcornmix <popcornmix@gmail.com>
Date: Wed, 16 May 2012 20:16:43 +0100
Subject: [PATCH] Add __VCCOREVER__ to makefile


diff --git a/sound/arm/Makefile b/sound/arm/Makefile
index 3939156..181cb57 100644
--- a/sound/arm/Makefile
+++ b/sound/arm/Makefile
@@ -18,5 +18,5 @@ snd-pxa2xx-ac97-objs		:= pxa2xx-ac97.o
 obj-$(CONFIG_SND_BCM2835)	+= snd-bcm2835.o
 snd-bcm2835-objs		:= bcm2835.o bcm2835-ctl.o bcm2835-pcm.o bcm2835-vchiq.o
 
-EXTRA_CFLAGS += -Idrivers/misc/vc04_services -Idrivers/misc/vc04_services/interface/vcos/linuxkernel
+EXTRA_CFLAGS += -Idrivers/misc/vc04_services -Idrivers/misc/vc04_services/interface/vcos/linuxkernel -D__VCCOREVER__=0x04000000
 
