From 07800e787edbea6f1c5d65e294b8cae85e81db76 Mon Sep 17 00:00:00 2001
From: popcornmix <popcornmix@gmail.com>
Date: Sat, 26 May 2012 16:40:51 +0100
Subject: [PATCH] Added power off message to allow kexec to work. Thanks lp0


diff --git a/arch/arm/mach-bcm2708/power.c b/arch/arm/mach-bcm2708/power.c
index 0c08678..d147bb6 100644
--- a/arch/arm/mach-bcm2708/power.c
+++ b/arch/arm/mach-bcm2708/power.c
@@ -161,6 +161,7 @@ static int __init bcm_power_init(void)
 	int i;
 
 	printk(KERN_INFO "bcm_power: Broadcom power driver\n");
+	bcm_mailbox_write(MBOX_CHAN_POWER, 0);
 
 	for (i = 0; i < BCM_POWER_MAXCLIENTS; i++)
 		g_state.client_request[i] = BCM_POWER_NOCLIENT;
