From 6273fcbf72461c2bd3757a8b6c9ed175955b25c1 Mon Sep 17 00:00:00 2001
From: popcornmix <popcornmix@gmail.com>
Date: Sat, 16 Jun 2012 23:46:08 +0100
Subject: [PATCH] Fix bool/int error, part 2


diff --git a/drivers/mmc/host/sdhci-bcm2708.c b/drivers/mmc/host/sdhci-bcm2708.c
index eedd724..13abbc9 100644
--- a/drivers/mmc/host/sdhci-bcm2708.c
+++ b/drivers/mmc/host/sdhci-bcm2708.c
@@ -136,7 +136,7 @@ static inline unsigned long int since_ns(hptime_t t)
 }
 
 static bool allow_highspeed = 1;
-static bool emmc_clock_freq = BCM2708_EMMC_CLOCK_FREQ;
+static int emmc_clock_freq = BCM2708_EMMC_CLOCK_FREQ;
 
 #if 0
 static void hptime_test(void)
